packages/vats/src/transfer.js

```diff
@@ -0,0 +1,242 @@
+              ackMethodData.ack = obj.acknowledgement;
+              retP = watch(E(targetHost).sendDowncall(ackMethodData));
+            }
+            retP = watch(retP, this.facets.sendErrorWatcher, { ackMethodData });
+          }
+
+          // Log errors in the upcall handling.
+          retP = watch(retP, this.facets.logErrorWatcher, { obj });
+
+          if (isActiveTap) {
+            // If the tap is active, return the promiseVow to the caller.  This
+            // will delay the middleware until fulfilment of the retP chain.
+            return retP;
+          }
+
+          // Otherwise, passively return nothing.
+        },
+      },
+      /**
+       * `watch` callback for encoding the raw `ack` return value from an active
+       * `writeAcknowledgement` tap as base64, then sending down to the
+       * targetHost.
+       */
+      encodeAckWatcher: {
+        onFulfilled(rawAck, { ackMethodData }) {
+          // Encode the tap's ack and write it out.
+          const ack = byteSourceToBase64(coerceToByteSource(rawAck));
+          ackMethodData = { ...ackMethodData, ack };
+          return E(this.state.targetHost).sendDowncall(ackMethodData);
+        },
+      },
+      /**
+       * `watch` callback for handling errors in the sending of an ack and
+       * reifying it as an error acknowledgement.
+       */
+      sendErrorWatcher: {
+        onRejected(error, { ackMethodData }) {
+          console.error(`Error sending ack:`, error);
+          const rawAck = JSON.stringify({ error: error.message });
+          ackMethodData = { ...ackMethodData, ack: byteSourceToBase64(rawAck) };
+          return E(this.state.targetHost).sendDowncall(ackMethodData);
+        },
+      },
+      /** `watch` callback for logging errors in the upcall handling. */
+      logErrorWatcher: {
+        onRejected(error, { obj }) {
+          console.error(`Error in handling of`, obj, error);
+        },
+      },
+    },
+  );
+
+  /**
+   * A TransferInterceptor wraps a TargetApp that is specialized for handling
+   * asynchronous callbacks from the IBC transfer protocol as implemented by
+   * `x/vtransfer`. By using this interceptor, the TargetApp is prevented from
+   * sending messages which may violate the transfer protocol directly to the
+   * `targetHost`.
+   *
+   * @param {Parameters<typeof makeTransferInterceptorKit>} args
+   */
+  const makeTransferInterceptor = (...args) =>
+    makeTransferInterceptorKit(...args).public;
+  return makeTransferInterceptor;
+};
+harden(prepareTransferInterceptor);
+
+const TransferMiddlewareI = M.interface('TransferMiddleware', {
+  registerTap: M.callWhen(
+    M.string(),
+    M.remotable('TransferInterceptor'),
+  ).returns(M.remotable('TargetRegistration')),
+  registerActiveTap: M.callWhen(
+    M.string(),
+    M.remotable('TransferInterceptor'),
+  ).returns(M.remotable('TargetRegistration')),
+  unregisterTap: M.callWhen(M.string()).returns(),
+});
```

Comment:
P.S. Is it intentional that TransferMiddleware supports register and unregister operations but not atomic reregister?

Reply:
No, that was an unintentional omission.